libkbfs/cr_chains.go

```diff
@@ -144,18 +145,81 @@ func (cc *crChain) getActionsToMerge(renamer ConflictRenamer, mergedPath path,
 }
 
 func (cc *crChain) isFile() bool {
+	return cc.file
+}
+
+// identifyType figures out whether this chain represents a file or
+// directory.  It tries to figure it out based purely on operation
+// state, but setAttr(mtime) can apply to either type; in that case,
+// we need to fetch the block to figure out the type.
+func (cc *crChain) identifyType(ctx context.Context, fbo *folderBlockOps,
+	md *RootMetadata, chains *crChains) error {
 	if len(cc.ops) == 0 {
-		return false
+		return nil
 	}
 
-	// If the first op is setAttr or sync, this is a file chain.
-	switch cc.ops[0].(type) {
-	case *syncOp:
-		return true
-	case *setAttrOp:
-		return true
+	// If any op is setAttr (ex or size) or sync, this is a file
+	// chain.  If it only has a setAttr/mtime, we don't know what it
+	// is, so fall through and fetch the block unless we come across
+	// another op that can determine the type.
+	var parentDir BlockPointer
+	for _, op := range cc.ops {
+		switch realOp := op.(type) {
+		case *syncOp:
+			cc.file = true
+			return nil
+		case *setAttrOp:
+			if realOp.Attr != mtimeAttr {
+				cc.file = true
+				return nil
+			}
+			// We can't tell the file type from an mtimeAttr, so we
+			// may have to actually fetch the block to figure it out.
+			parentDir = realOp.Dir.Ref
+		default:
+			return nil
+		}
+	}
+
+	parentOriginal, ok := chains.originals[parentDir]
+	if !ok {
+		return NoChainFoundError{parentDir}
 	}
-	return false
+
+	// We have to find the current parent directory block.  If the
+	// file has been renamed, that might be different from parentDir
+	// above.
+	if newParent, _, ok := chains.renamedParentAndName(cc.original); ok {
+		parentOriginal = newParent
+	}
+
+	parentMostRecent, err := chains.mostRecentFromOriginalOrSame(parentOriginal)
+	if err != nil {
+		return err
+	}
+
+	// If we get down here, we have an ambiguity, and need to fetch
+	// the block to figure out the file type.
+	dblock, err := fbo.GetDirBlockForReading(ctx, makeFBOLockState(), md,
+		parentMostRecent, fbo.folderBranch.Branch, path{})
+	if err != nil {
+		return err
+	}
+	// We don't have the file name handy, so search for the pointer.
+	found := false
+	for _, entry := range dblock.Children {
+		if entry.BlockPointer != cc.mostRecent {
+			continue
+		}
+		cc.file = entry.Type != Dir
```

Comment:
Maybe we should switch on `entry.Type` and return an error if it's something other than `File`, `Dir`, or `Exec`.